text/0062-consistent-behavior-for-video-sources-that-do-not-have-video.md

```diff
@@ -0,0 +1,90 @@
+Some video sources, like window capture, video capture, or media sources, at some point in their lifetime might not have any video to display.
+Currently, what sources do is that scenario is pretty much a wild west.
+For example, *macOS Screen Capture* shows the last known frame, *Media Sources* have a checkbox between invisible and the last known frame, *Game Capture* (Windows) shows nothing (to my knowledge), etc.
+Past discussions have shown that different users expect differnt things to happen in this case.
+Some want the last frame, others want nothing, etc.
+Part of this are privacy concerns on what is shown, however this goes both ways: A window capture that shows the last frame may not be supposed to do that as the user clicked the window away since they showed something they didn't mean to; meanwhile a source that suddenly shows nothing could expose what is below it.
+Giving control over this to the user should address those conerns.
+
+## Details
+### New APIs / Internals
+The problem with the current situation is that every source needs to implement its own property to show basically the same thing.
+To simplify this, the idea is to not use the source properties/settings API.
+We create a setting that's not stored as part of the usual source settings and can be queried by the source.
+
+A video source that at some point might not have video it should indicate so via a new capability flag (The exact value of this might change if (1 << 17) is taken by the time of implementation) in its `obs_source_info`:
+```c
+#define OBS_SOURCE_CAP_MIGHT_NOT_HAVE_VIDEO (1 << 17)
+```
+Setting this flag requires that the source implements and respects the other API's shown below.
+Any source that might not have video should set this.
+The idea of a new flag is that not all sources (e.g., text sources or color sources) require such a setting, so they should not get a UI for this or need to implement new APIs.
+
+
+We create an enum to for the available states, as well as methods to get this from and set it on the source.
+```c
+enum obs_source_no_video_behavior {
+    OBS_SOURCE_NO_VIDEO_BEHAVIOR_SHOW_BLACK_FRAME,
+    OBS_SOURCE_NO_VIDEO_BEHAVIOR_SHOW_NOTHING,
+    OBS_SOURCE_NO_VIDEO_BEHAVIOR_SHOW_LAST_FRAME,
+}
+enum obs_source_no_video_behavior obs_source_get_no_video_behavior(obs_source_t *source);
+void obs_source_set_no_video_behavior(obs_source_t *source, enum obs_source_no_video_behavior behavior);
+```
+If a source at some point has no video, it should query what state is set via
+`obs_source_get_no_video_behavior`
+According to what is set, it is then the source's responsibility to output what is asked of it.
+For this, sources should remember the last frame and its dimensions.
+The value of this can change at any time (for the avoidance of doubt, this includes when there currently is no video), and should always be respected.
+If a source is supposed to show the last frame but never had one, it should show a black frame instead.
```

Comment:
I just realized I didn't completely think this line through - how is the source meant to have a resolution? Make a best guess? Hardcoded 1080x1920? Will need more thought.